src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs

```diff
@@ -21,115 +19,148 @@ namespace Microsoft.CodeAnalysis.Remote
+                listener,
+                shutdownToken);
 
             // start listening workspace change event
             _workspace.WorkspaceChanged += OnWorkspaceChanged;
+            _globalOperationService.Started += OnGlobalOperationStarted;
+            _globalOperationService.Stopped += OnGlobalOperationStopped;
 
-            // create its own cancellation token source
-            _globalOperationCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(shutdownToken);
-
-            Start();
+            // Enqueue the work to sync the initial solution.
+            ResumeWork();
         }
 
-        protected override async Task ExecuteAsync()
+        public void Shutdown()
         {
-            lock (_gate)
-            {
-                Contract.ThrowIfNull(_currentToken);
-                _currentToken.Dispose();
-                _currentToken = null;
-            }
+            // Try to stop any work that is in progress.
+            PauseWork();
+            _pauseCancellationSeries.Dispose();
 
-            // update primary solution in remote host
-            await SynchronizePrimaryWorkspaceAsync(_globalOperationCancellationSource.Token).ConfigureAwait(false);
+            _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+            _globalOperationService.Started -= OnGlobalOperationStarted;
+            _globalOperationService.Stopped -= OnGlobalOperationStopped;
         }
 
-        protected override void OnPaused()
-        {
-            var previousCancellationSource = _globalOperationCancellationSource;
-
-            // create new cancellation token source linked with given shutdown cancellation token
-            _globalOperationCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(this.CancellationToken);
+        private void OnGlobalOperationStarted(object? sender, EventArgs e)
+            => PauseWork();
 
-            CancelAndDispose(previousCancellationSource);
-        }
+        private void OnGlobalOperationStopped(object? sender, EventArgs e)
+            => ResumeWork();
 
-        protected override async Task WaitAsync(CancellationToken cancellationToken)
+        private void PauseWork()
         {
-            var currentToken = await _workQueue.DequeueAsync(cancellationToken).ConfigureAwait(false);
-            lock (_gate)
-            {
-                Contract.ThrowIfFalse(_currentToken is null);
-                _currentToken = currentToken;
-            }
+            // An expensive global operation started (like a build).  Pause ourselves and cancel any outstanding work in
+            // progress to synchronize the solution.
+            //
+            // Note: We purposefully ignore the cancellation token produced by CreateNext.  The purpose here is to
+            // cancel the token controlling the current work, but not have any uncanceled token for new work to use when
+            // it comes in.
```

Comment:
Did you get a sense if this really does much? I have a internal perf bug that was hinting that we might be doing lots of sync operations when we are competing against other work. But even this means we're not _preemptively_ doing a sync, any individual feature could still force the sync to happen anyways in response to any other operation, correct?